extensions/kinesis/src/main/java/com/hazelcast/jet/kinesis/impl/ShardReader.java

```diff
@@ -0,0 +1,278 @@
+import com.amazonaws.services.kinesis.model.GetShardIteratorResult;
+import com.amazonaws.services.kinesis.model.ProvisionedThroughputExceededException;
+import com.amazonaws.services.kinesis.model.Record;
+import com.amazonaws.services.kinesis.model.Shard;
+import com.hazelcast.logging.ILogger;
+
+import java.util.LinkedList;
+import java.util.concurrent.Future;
+
+import static com.hazelcast.jet.impl.util.ExceptionUtil.rethrow;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+class ShardReader extends AbstractShardWorker {
+
+    /* Kinesis allows for a maximum of 5 GetRecords operations per second. */
+    private static final int GET_RECORD_OPS_PER_SECOND = 5;
+
+    /* Even though GetRecords operations are limited to 5 per second, if one
+     * such operation happens to return too much data, following operations will
+     * throw ProvisionedThroughputExceededException. In such cases we need to
+     * wait a bit longer than for regular rate limiting.
+     *
+     * Relevant section from AWS documentation:
+     *
+     * "The size of the data returned by GetRecords varies depending on the
+     * utilization of the shard. The maximum size of data that GetRecords can
+     * return is 10 MiB. If a call returns this amount of data, subsequent calls
+     * made within the next 5 seconds throw
+     * ProvisionedThroughputExceededException. If there is insufficient
+     * provisioned throughput on the stream, subsequent calls made within the
+     * next 1 second throw ProvisionedThroughputExceededException. GetRecords
+     * doesn't return any data when it throws an exception. For this reason, we
+     * recommend that you wait 1 second between calls to GetRecords. However,
+     * it's possible that the application will get exceptions for longer than
+     * 1 second."
+     *
+     * We also need to add this extra wait whenever we encounter other unexpected
+     * failures.
+     * */
+    private static final long PAUSE_AFTER_FAILURE = SECONDS.toNanos(1); //todo: exponential backoff
+
+    /**
+     * Maximum number of records returned by this reader in a single batch. Is
+     * limited due to being used from a cooperative processor. Should not pose
+     * a performance bottleneck, because while available data is being processed
+     * the asynchronous request for more will already be issued in the background.
+     */
+    private static final int DATA_BATCH_SIZE = 100;
+
+    private final Shard shard;
+    private final RandomizedRateTracker getRecordsRateTracker =
+            new RandomizedRateTracker(1000, GET_RECORD_OPS_PER_SECOND);
+
+    private State state = State.NO_SHARD_ITERATOR;
+    private String shardIterator;
+    private Future<GetShardIteratorResult> shardIteratorResult;
+    private Future<GetRecordsResult> recordsResult;
+    private long nextGetRecordsTime = System.nanoTime();
+
+    private final LinkedList<Record> data = new LinkedList<>();
```

Comment:
What about using `ArrayDeque`? The linked list uses more memory and has more dereferencing. It allows for random removals, but we don't need that, we only remove from one and and add to the other.